analyzers/src/SonarAnalyzer.CFG/Roslyn/CfgAllPathValidator.cs

```diff
@@ -37,16 +44,32 @@ public abstract class CfgAllPathValidator
         public bool CheckAllPaths() =>
             IsBlockOrAllSuccessorsValid(this.cfg.EntryBlock);
 
-        private bool IsBlockOrAllSuccessorsValid(BasicBlock block) =>
-            !IsInvalid(block) && (IsValid(block) || AreAllSuccessorsValid(block));
+        private bool IsBlockOrAllSuccessorsValid(BasicBlock block)
+        {
+            var isValid = !IsInvalid(block) && (IsValid(block) || AreAllSuccessorsValid(block));
+            lattice[block] = isValid ? BlockResult.TRUE : BlockResult.FALSE;
+            return isValid;
+        }
 
         private bool AreAllSuccessorsValid(BasicBlock block)
         {
-            this.visited.Add(block);
-
-            return !block.SuccessorBlocks.Contains(cfg.ExitBlock)
-                   && block.SuccessorBlocks.Except(visited).Any()
-                   && block.SuccessorBlocks.Except(visited).All(IsBlockOrAllSuccessorsValid);
+            lattice[block] = BlockResult.INCONCLUSIVE;
+            if (!block.SuccessorBlocks.Contains(cfg.ExitBlock) && block.SuccessorBlocks.Any())
```

Comment:
Let's swap this. I find it easier to understand.
```suggestion
            if (block.SuccessorBlocks.Any() && !block.SuccessorBlocks.Contains(cfg.ExitBlock))
```